crates/wallet/src/lib.rs

```diff
@@ -2,46 +2,214 @@
-//!   code.
+//! - `odyssey_sendTransaction` that can perform service-sponsored [EIP-7702][eip-7702] delegations
+//!   and send other service-sponsored transactions on behalf of EOAs with delegated code.
 //!
 //! # Restrictions
 //!
 //! `odyssey_sendTransaction` has additional verifications in place to prevent some
-//! rudimentary abuse of the sequencer's funds. For example, transactions cannot contain any
+//! rudimentary abuse of the service's funds. For example, transactions cannot contain any
 //! `value`.
 //!
 //! [eip-5792]: https://eips.ethereum.org/EIPS/eip-5792
 //! [eip-7702]: https://eips.ethereum.org/EIPS/eip-7702
 
 #![cfg_attr(not(test), warn(unused_crate_dependencies))]
 
-use alloy_eips::BlockId;
 use alloy_network::{
-    eip2718::Encodable2718, Ethereum, EthereumWallet, NetworkWallet, TransactionBuilder,
+    eip2718::Encodable2718, Ethereum, EthereumWallet, Network, NetworkWallet, TransactionBuilder,
 };
-use alloy_primitives::{Address, ChainId, TxHash, TxKind, U256};
-use alloy_rpc_types::TransactionRequest;
+use alloy_primitives::{Address, Bytes, ChainId, TxHash, TxKind, U256};
+use alloy_provider::{utils::Eip1559Estimation, Provider, WalletProvider};
+use alloy_rpc_types::{BlockId, TransactionRequest};
+use alloy_transport::Transport;
 use jsonrpsee::{
     core::{async_trait, RpcResult},
     proc_macros::rpc,
 };
 use metrics::Counter;
 use metrics_derive::Metrics;
+
 use reth_rpc_eth_api::helpers::{EthCall, EthTransactions, FullEthApi, LoadFee, LoadState};
-use reth_storage_api::{StateProvider, StateProviderFactory};
-use revm_primitives::Bytecode;
+use reth_storage_api::StateProviderFactory;
 use serde::{Deserialize, Serialize};
-use std::sync::Arc;
+use std::{marker::PhantomData, sync::Arc};
 use tracing::{trace, warn};
 
 use reth_optimism_rpc as _;
 use tokio::sync::Mutex;
 
-/// The capability to perform [EIP-7702][eip-7702] delegations, sponsored by the sequencer.
+/// A trait for any type capable of estimating, signing, and propagating sponsored transactions.
+#[async_trait]
+pub trait Node<N>
+where
+    N: Network,
+{
+    /// Get the address of the account that sponsors transactions.
+    fn default_signer_address(&self) -> Address;
+
+    /// Get the code at a specific address.
+    async fn get_code(&self, address: Address) -> Result<Bytes, OdysseyWalletError>;
+
+    /// Estimate the transaction request's gas usage and fees.
+    async fn estimate(
+        &self,
+        tx: &N::TransactionRequest,
+    ) -> Result<(u64, Eip1559Estimation), OdysseyWalletError>;
+
+    /// Sign the transaction request and send it to the node.
+    async fn sign_and_send(&self, tx: N::TransactionRequest) -> Result<TxHash, OdysseyWalletError>;
+}
+
+/// A wrapper around an Alloy provider for signing and sending sponsored transactions.
+#[derive(Debug)]
+pub struct AlloyNode<P, T, N> {
+    provider: P,
+    _transport: PhantomData<T>,
+    _network: PhantomData<N>,
+}
```

Comment:
do we expect that we need more than just the alloy provider type?

if not we don't need this and can impl the trait on Provider directly?

Reply:
it complains about T being unconstrained in that case, I've tried.